rust_snuba/src/processors/generic_metrics.rs

```diff
@@ -0,0 +1,1138 @@
+use adler32::adler32;
+use anyhow::Context;
+use chrono::DateTime;
+use rust_arroyo::backends::kafka::types::KafkaPayload;
+use serde::{Deserialize, Serialize};
+use std::{collections::BTreeMap, vec};
+
+use crate::{
+    types::{InsertBatch, RowData},
+    KafkaMessageMetadata, ProcessorConfig,
+};
+
+use super::utils::enforce_retention;
+
+const GRANULARITY_TEN_SECONDS: u8 = 0;
+const GRANULARITY_ONE_MINUTE: u8 = 1;
+const GRANULARITY_ONE_HOUR: u8 = 2;
+const GRANULARITY_ONE_DAY: u8 = 3;
+
+/// Generate a timeseries ID from the given parameters. Timeseries IDs are used to
+/// uniquely identify a timeseries in the database. This implemenation is based on
+/// the python implementation in order to ensure that the same timeseries ID is
+/// generated for the same input.
+///
+/// Upstream sends the tag keys and values as strings. Which is surprising since the
+/// tag keys are indexed so they should be integers. But we have to deal with it.
+fn generate_timeseries_id(
+    org_id: u64,
+    project_id: u64,
+    metric_id: u64,
+    tags: &BTreeMap<String, String>,
+) -> u32 {
+    let mut buffer = Vec::new();
+    buffer.extend_from_slice(&org_id.to_le_bytes());
+    buffer.extend_from_slice(&project_id.to_le_bytes());
+    buffer.extend_from_slice(&metric_id.to_le_bytes());
+
+    for (key, value) in tags {
+        buffer.extend_from_slice(key.as_bytes());
+        buffer.extend_from_slice(value.as_bytes());
+    }
+
+    adler32(&*buffer).unwrap()
+}
+
+#[derive(Debug, Deserialize)]
+struct FromGenericMetricsMessage {
+    use_case_id: String,
+    org_id: u64,
+    project_id: u64,
+    metric_id: u64,
+    #[serde(rename = "type")]
+    r#type: String,
+    timestamp: u64,
+    sentry_received_timestamp: SentryReceivedTimestamp,
+    tags: BTreeMap<String, String>,
+    value: MetricValue,
+    retention_days: u16,
+    aggregation_option: Option<String>,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(untagged)]
+enum MetricValue {
+    Counter(f64),
+    SetOrDistribution(Vec<u64>),
+    DistributionFloat(Vec<f64>),
+    Gauge {
+        count: u64,
+        last: f64,
+        max: f64,
+        min: f64,
+        sum: f64,
+    },
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(untagged)]
+enum SentryReceivedTimestamp {
```

Comment:
My initial thought was that we would only receive u64. But Lyn correctly pointed out that the schema examples had a float in there. Reference here https://github.com/getsentry/snuba/pull/5360#discussion_r1446774263

Reply:
You only need to parse a `f64` here. JSON doesn't really have an integer anyway, everything is a float, so it works.

Reply:
changed it to f64

Reply:
I now tested it locally, and indeed serde deserializes integers into floats without error.

I wasn't sure because Python certainly distinguishes between integer vs float:

```
>>> import json
>>> json.loads("123.0")
123.0
>>> json.loads("123")
123
>>>
```